crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -374,60 +380,122 @@ pub fn extract_skinned_meshes(
     commands.insert_or_spawn_batch(values);
 }
 
+/// Data necessary to be equal for two draw commands to be mergeable
+///
+/// This is based on the following assumptions:
+/// - Only entities with prepared assets (pipelines, materials, meshes) are
+///   queued to phases
+/// - View bindings are constant across a phase for a given draw function as
+///   phases are per-view
+/// - `prepare_mesh_uniforms` is the only system that performs this batching
+///   and has sole responsibility for preparing the per-object data. As such
+///   the mesh binding and dynamic offsets are assumed to only be variable as a
+///   result of the `prepare_mesh_uniforms` system, e.g. due to having to split
+///   data across separate uniform bindings within the same buffer due to the
+///   maximum uniform buffer binding size.
+#[derive(PartialEq, Eq)]
+struct BatchMeta3d {
+    /// The pipeline id encompasses all pipeline configuration including vertex
+    /// buffers and layouts, shaders and their specializations, bind group
+    /// layouts, etc.
+    pipeline_id: CachedRenderPipelineId,
+    /// The draw function id defines the RenderCommands that are called to
+    /// set the pipeline and bindings, and make the draw command
+    draw_function_id: DrawFunctionId,
+    material_bind_group_id: Option<MaterialBindGroupId>,
+    mesh_asset_id: AssetId<Mesh>,
+    mesh_flags: u32,
+    dynamic_offset: Option<NonMaxU32>,
+}
+
+impl BatchMeta<BatchMeta3d> for BatchMeta3d {
+    #[inline]
+    fn matches(&self, other: &BatchMeta3d) -> bool {
+        self.pipeline_id == other.pipeline_id
+            && self.draw_function_id == other.draw_function_id
+            && self.mesh_asset_id == other.mesh_asset_id
+            && (self.mesh_flags & (MeshFlags::SKINNED | MeshFlags::MORPH_TARGETS).bits()) == 0
```

Comment:
i understand the logic here now. but i think we also need to check `other.mesh_flags`, else we would count the first skinned/morphed mesh as also part of the previous batch

Reply:
I may be wrong, but I don't think so. This checks whether the skinned or morph targets flags are set, and if so it returns false. That is, if this entity is skinned or has morph targets, it cannot be batched with anything. I can add a comment to that effect.

As for checking mesh flags... it depends on the impact of the mesh flags. If they require splitting batches or not depends a lot on the flag. So for now I think it has to be case-by-case, but again, longer term I really want the batching to be based on things that would actually require separate draws due to changes to any bindings.

Reply:
you're right, your batching code checks for `next.matches(current)` so it's fine. i changed it on my pr as there i think i check `current == next`

Reply:
And when I looked at your PR, I think you are also right. It shouldn't be sensitive to `current == next` or `next == current` ordering.

Reply:
maybe morph and skinning could insert `NoAutomaticBatching` instead of using the bits? that would be more extensible.

also should `NoAutomaticBatching` use SparseSet storage? i would hope the perf cost for a query using `Without` on a sparse set component would be zero (vs a table component).

Reply:
Hmm, using `NoAutomaticBatching` for morph targets and skinning are an option. I'm not sure though. Whenever I've used sparse storage with the ECS, iteration performance has fallen over. Regardless, it would have to be kept in sync with the truth. I guess the flags have to as well.

Reply:
Done.